mesonbuild/coredata.py

```diff
@@ -1069,6 +1069,33 @@ def parse_cmd_line_options(args: argparse.Namespace) -> None:
             args.cmd_line_options[key] = value
             delattr(args, name)
 
+def validate_dirs(builddir: T.Optional[str], builddir_or_sourcedir: T.Optional[str]) -> T.Tuple[T.Optional[Path], T.Optional[Path]]:
+    if builddir is not None:
+        if builddir_or_sourcedir is not None:
+            raise MesonException('builddir defined with -C argument and positional argument')
+        build_or_source = builddir
+    elif builddir_or_sourcedir is None:
+        build_or_source = '.'
+    else:
+        build_or_source = builddir_or_sourcedir
+
+    build_or_source = Path(build_or_source).resolve()
+    if Path(build_or_source, 'meson-private').exists():
+        builddir = build_or_source
+        sourcedir = None
+    elif builddir is not None:
+        raise MesonException(f'{builddir} is not a Meson build directory.')
```

Comment:
It's a pity that "Please specify a valid build dir or change the working directory to it." is lost.

Reply:
That message was also slightly wrong because you could point to a source directory too. I updated the message now.